packages/controllers/src/permissions/PermissionController.ts

```diff
@@ -0,0 +1,1970 @@
+    originalMetadata: PermissionsRequestMetadata,
+  ) {
+    const { id, origin } = originalMetadata;
+
+    if (
+      !isPlainObject(approvedRequest) ||
+      !isPlainObject(approvedRequest.metadata)
+    ) {
+      throw internalError(
+        `Approved permissions request for subject "${origin}" is invalid.`,
+        { data: { approvedRequest } },
+      );
+    }
+
+    const {
+      metadata: { id: newId, origin: newOrigin },
+      permissions,
+    } = approvedRequest;
+
+    if (newId !== id) {
+      throw internalError(
+        `Approved permissions request for subject "${origin}" mutated its id.`,
+        { originalId: id, mutatedId: newId },
+      );
+    }
+
+    if (newOrigin !== origin) {
+      throw internalError(
+        `Approved permissions request for subject "${origin}" mutated its origin.`,
+        { originalOrigin: origin, mutatedOrigin: newOrigin },
+      );
+    }
+
+    try {
+      this.validateRequestedPermissions(origin, permissions);
+    } catch (error) {
+      // Re-throw as an internal error; we should never receive invalid approved
+      // permissions.
+      throw internalError(
+        `Invalid approved permissions request: ${error.message}`,
+        error.data,
+      );
+    }
+  }
+
+  /**
+   * Accepts a permissions request created by
+   * {@link PermissionController.requestPermissions}.
+   *
+   * @param request - The permissions request.
+   */
+  async acceptPermissionsRequest(request: PermissionsRequest): Promise<void> {
+    const { id } = request.metadata;
+
+    if (!this.hasApprovalRequest({ id })) {
+      throw new PermissionsRequestNotFoundError(id);
+    }
+
+    if (Object.keys(request.permissions).length === 0) {
+      this._rejectPermissionsRequest(
+        id,
+        invalidParams({
+          message: 'Must request at least one permission.',
+        }),
+      );
+      return;
+    }
+
+    try {
+      this.messagingSystem.call(
+        'ApprovalController:acceptRequest',
+        id,
+        request,
+      );
+    } catch (error) {
+      // If accepting unexpectedly fails, reject the request and re-throw the
+      // error
+      this._rejectPermissionsRequest(id, error);
+      throw error;
+    }
+  }
+
+  /**
+   * Rejects a permissions request created by
+   * {@link PermissionController.requestPermissions}.
+   *
+   * @param id - The id of the request to be rejected.
+   */
+  async rejectPermissionsRequest(id: string): Promise<void> {
+    if (!this.hasApprovalRequest({ id })) {
+      throw new PermissionsRequestNotFoundError(id);
+    }
+
+    this._rejectPermissionsRequest(id, userRejectedRequest());
+  }
+
+  /**
+   * Checks whether the {@link ApprovalController} has a particular permissions
+   * request.
+   *
+   * @see {@link PermissionController.acceptPermissionsRequest} and
+   * {@link PermissionController.rejectPermissionsRequest} for usage.
+   * @param options - The {@link HasApprovalRequest} options.
```

Comment:
Nit: we ought to document the `id` option here as well

Reply:
91cf3df